sbpy/photometry/core.py

```diff
@@ -99,24 +104,306 @@ def mag2ref(mag, radius, M_sun=None):
+
+    ``SpectralGradient`` is a `~astropy.units.Quantity` object with
+    units of percent change per wavelength.
+
+    Spectral gradient is with respect to the flux density at a
+    particular wavelength.  By convention this is the mean flux
+    between the two filters, assuming the reflectance spectrum is
+    linear with wavelength.  Eq. 1 of A'Hearn et al. [ADT84]_:
+
+    .. math::
+
+        S = \frac{R(λ1) - R(λ0)}{R(λ1) + R(λ0)} \frac{2}{Δλ}
+
+    Δλ is typically measured in units of 100 nm.
+
+
+    Parameters
+    ----------
+    value : number, `~astropy.units.Quantity`
+        The value(s).
+
+    unit : string, `~astropy.units.Unit`, optional
+        The unit of the input value.  Strings must be parseable by
+        :mod:`~astropy.units` package.
+
+    wave : `~astropy.units.Quantity`, optional
+        Effective wavelengths of the measurement for a solar spectral
+        energy distribution.  Required for conversion to other
+        wavelengths.
+
+    wave0 : `~astropy.units.Quantity`, optional
+        Normalization point.  If ``None``, the mean of ``wave``
+        will be assumed.
+
+    dtype : `~numpy.dtype`, optional
+        See `~astropy.units.Quantity`.
+
+    copy : bool, optional
+        See `~astropy.units.Quantity`.
+
+
+    Examples
+    --------
+    >>> import astropy.units as u
+    >>> from sbpy.units import hundred_nm
+    >>> S = SpectralGradient(10 * u.percent / hundred_nm, wave0=5500 * u.AA)
+    >>> print(S)
+    10.0 % / 100 nm
+
+    >>> from sbpy.units import VEGAmag
+    >>> bp = ('johnson_v', 'cousins_r')
+    >>> VmR = 15.8 * VEGAmag - 15.3 * VEGAmag
+    >>> VmR_sun = 0.37 * u.mag
+    >>> S = SpectralGradient.from_color(bp, VmR - VmR_sun)
+    ...                              # doctest: +REMOTE_DATA +IGNORE_OUTPUT
+    >>> print(S)                     # doctest: +REMOTE_DATA +FLOAT_CMP
+    12.29185986266534 % / 100 nm
+
+
+    References
+    ----------
+    .. [ADT84] A'Hearn, Dwek & Tokunaga 1984. Infrared Photometry of
+        Comet Bowell and Other Comets. ApJ 282, 803-806.
+
+    """
+
+    _equivalent_unit = u.meter**-1
+    _include_easy_conversion_members = False
+
+    def __new__(cls, value, unit=None, wave=None, wave0=None,
+                dtype=None, copy=None):
+        S = super().__new__(cls, value, unit=unit, dtype=dtype,
+                            copy=copy)
+
+        if wave is not None:
+            if np.size(wave) != 2:
+                raise ValueError(
+                    'Two wavelengths must be provided, got {}'
+                    .format(np.size(wave)))
+            S.wave = S._eff_wave(wave)
+
+        if wave0 is None and wave is not None:
+            S.wave0 = S.wave.mean()
+        else:
+            S.wave0 = wave0
+
+        return S
+
+    @classmethod
+    def _eff_wave(cls, wfb):
+        """Wavelength/frequency/bandpass to wavelength.
+
+        Bandpass is converted to effective wavelength using a solar
+        spectrum.
```

Comment:
Should the option be open for a reference spectrum other than solar when calculating effective wavelength from bandpass?  Might be useful for extremely red object, but I don't know how of difference might be, and therefore the necessity.  Or maybe we can redden the solar spectrum based on `S.value` for effective wavelength calculation?

Reply:
Oh, that'll be worth investigation to see how large an effect it could be.  I'm fine with an iteration, but it would have to assume a functional form of the reddening (probably linear).  Maybe an optional iteration?

Reply:
Optional iteration is fine because it has the risk of slow and not converge.   Linear reddening is the only option we can take I guess.

Reply:
We should make a reddening function, or look into synphot for what is used there.  Since we agree that this should be optional, I recommend we save it for v0.2.

Reply:
Sounds good to me.